### PR TITLE
Add `predicates` to allow filtering `watcher` streams

### PR DESCRIPTION
**Pull request by clux**
A re-implementation of the [go predicates from controller-runtime](https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/predicate) that allows filtering `watcher` streams from a new (unstable) `WatchStreamExt::predicate_filter`:

```rust
let changed_pods = watcher(pods, watcher::Config::default())
    .applied_objects()
    .predicate_filter(predicates::generation);

while let Some(pod) = changed_pods.try_next().await? {
   info!("saw Pod '{} with hitherto unseen generation", pod.name_any());
}
```

This allows filtering out events that only changes (say) the status object. It contains a set of common extractor functions (under `runtime::predicates`) and operates through `PredicateFilter`. This stream utility contains a small `HashMap` with hashes of the last seen value for each object, and simply compares that value (from the extractor) with the one in the new watch event.

<details><summary>notes and origin and discussion links</summary>
<p>
based on user predicates that allow short-circuiting `reconciler` runs if nothing important has changed:

see [this commit](https://github.com/kube-rs/controller-rs/commit/7e2a15274d0ef99a3b7c841446bc104535ed2118) which was used in the reconciler as an early exit gate via:

```rust
    if !generation.cmp_update(&mut *ctx.predicate_cache.write().await, &doc) {
        info!("ignoring generationally equivalent reconcile for {}", doc.name());
        return Ok(Action::requeue(Duration::from_secs(30 * 60)));
    }
```
relying on passing an thread-safe `Evaluation` cache via the `reconcile` context. This method has huge problems when used in controllers with more than one type, because this setup only filters based on the merged QueueStream. Crucially, if an owned object changes, it won't be captured in the root type, so you might miss important changes to child objects. Doing the check inside the reconciler is too late.

So for controller usage we have to do predicate style filters on the watcher streams instead, and have users configure predicates PER stream, then pass them to the controller using stream sharing for this to work cohesively.

- initial [discord thread](https://discord.com/channels/500028886025895936/974686542104772628)
- separate [discord thread](https://discord.com/channels/500028886025895936/1090650140282454037)

</p>
</details>

### CODE REVIEW

**Comment by clux** [comment]
As noted on discord this is problematic because:

1. **it only accounts for the last object in the controller's queue stream**
2. **it encourages non-idempotent modes of operation of a reconciler**

### First problem
I think it's possible to work around this by:

- add the filter at the watcher level (in `WatchStreamExt`)
- pass watcher streams into the controller directly

That way noop-considered change for every object is ignored and not propagated further inside the controller.
Even if we don't allow passing filtered watcher streams into controllers, it can still be useful to have this as a watcher filter.

### Second problem

More insidious. People are going to get repeat reconcile requests even when no changes were made, the system is meant to requeue based on time, and controller restarts triggers new reconciles, and users may want to manually trigger the reconciliation somehow without a functional change (that affects generation say), and a predicate system may ultimately inhibit that.

Would love to hear more thoughts on this, we talked about it on discord, but it was brief and in the middle of many different threads.

**Comment by clux** [comment]
Have reduced this in scope to not be about controllers. It's currently now just something we can use to filter flattened watcher streams:

```rust
let stream = watcher(node, lp).applied_objects().filter_predicates(predicates::annotations);
pin_mut!(stream);
while let Some(node) = rf.try_next().await? {
    info!("saw node {} with hitherto unseen annotations", node.name());
}
```

It does the filtering by maintaining a reasonably optimised cache (only storing hashed values - so limits predicate fns instead), which is important because objects coming and going at the usual kubernetes rate is going to leave this cache with lots of potential to grow. Ideally, the map should be some timelimited cache where keys drop out of scope after some time.

I think this use case still has value in this form outside controllers (where they are slightly more problematic), but probably not going to follow-up until next week due to KubeCon. Feedback welcome.

**Comment by clux** [comment]
Opening this up for review because I think this has value as is when plugged through `watcher` directly, and have removed/hidden references to controller ideas for it (which are a little more controversial).

**Comment by clux** [comment]
Wrapped this in an unstable feature and added better docs. Hoping to get this in so we can iterate if we find corner cases. I still have not found good ways to solve this, and it can actually be useful with something like #1163 as an early place to experiment with stream sharing with stuff like #1163. cc @Dav1dde 

Deny will fail because of a dev-dep duplicate (it's fixed in #1171)

File: kube-runtime/src/utils/predicate.rs [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+
+#[allow(clippy::pedantic)]
+#[pin_project]
+/// Stream returned by the [`predicate_filter`](super::WatchStreamExt::predicate_filter) method.
+#[must_use = "streams do nothing unless polled"]
+#[allow(clippy::type_complexity)]
+pub struct PredicateFilter<St, K: Resource> {
+    #[pin]
+    stream: St,
+    predicate: Box<dyn (Fn(&K) -> Option<u64>)>,
```
[the remark this answers is not in the document]
Comment (clux) [reply kube-runtime/src/utils/predicate.rs]: hm ok, i've added `Send` restrictions.

[the remark this answers is not in the document]
Comment (clux) [reply kube-runtime/src/utils/predicate.rs]: No particular reason. Is https://github.com/kube-rs/kube/pull/911/commits/894b72b7289e6a11235e089f63038580fd345f48 more like what you were thinking?

[the remark this answers is not in the document]
Comment (nightkr) [reply kube-runtime/src/utils/predicate.rs]: Pretty much, except we no longer need to box it, and we can remove the `Send` constraint (it'll be `Send` anyway if all of its contents are `Send`).

**Comment by clux** [comment]
Going to send this through under the unstable feature to let it stew for a while. This is a simplistic solution and may benefit from some TTL style decay (or moving it before the flatteners to get access to raw `watcher::Event`s) in the future. But, in the mean time, would like to run this in some of my controllers for a while to see how well it works, and if that type of over-engineering is necessary.

**Pull request merged**